File: code_samples/section3/problems/section3.py
"""Section 3 - Linked Lists"""

#pylint: disable=too-few-public-methods
#pylint: disable=redefined-builtin

from typing import Optional

class ListNode:
    """Linked list node."""
    def __init__(self, val: int = 0, next: "Optional[ListNode]" = None) -> None:
        self.val = val
        self.next = next

def insert_head(head: Optional[ListNode], x: int) -> ListNode:
    """Insert a value at the head of the linked list."""
    return ListNode(x, head)

def insert_after(p: Optional[ListNode], x: int) -> None:
    """Insert a value after a given node."""
    if p is None:
        return
    node = ListNode(x, p.next)
    p.next = node

def delete_after(p: Optional[ListNode]) -> None:
    """Delete a node after a given node."""
    if p is None or p.next is None:
        return
    doomed = p.next
    p.next = doomed.next

def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """Reverse a linked list."""
    prev: Optional[ListNode] = None
    curr = head
    while curr is not None:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

def has_cycle(head: Optional[ListNode]) -> bool:
    """Check if a linked list has a cycle."""
    if head is None:
        return False
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return True
    return False

def middle_node(head: Optional[ListNode]) -> Optional[ListNode]:
    """Find the middle node of a linked list."""
    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
    return slow

def merge_two_lists(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """Merge two sorted linked lists."""
    dummy = ListNode(0)
    tail = dummy

    a = l1
    b = l2
    while a is not None and b is not None:
        if a.val <= b.val:
            tail.next = a
            a = a.next
        else:
            tail.next = b
            b = b.next
        tail = tail.next
    if a is not None:
        tail.next = a
    if b is not None:
        tail.next = b

    return dummy.next

def remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """Remove the nth node from the end of a linked list."""
    dummy = ListNode(0, head)
    fast: Optional[ListNode] = dummy
    slow: Optional[ListNode] = dummy

    for _ in range(n + 1):
        if fast is None:
            return head
        fast = fast.next

    while fast is not None:
        fast = fast.next
        slow = slow.next  # type: ignore[assignment]

    doomed = slow.next  # type: ignore[assignment]
    if doomed is not None:
        slow.next = doomed.next  # type: ignore[assignment]
    return dummy.next

def print_list(head: "ListNode | None") -> None:
    """Print linked list."""
    current = head
    print()
    while current is not None:
        print(current.val, end="")
        current = current.next

def main() -> None:
    """Main function."""
    # ----- Build list: 1 2 3 4 5 -----
    head: Optional[ListNode] = None
    head = insert_head(head, 5)
    head = insert_head(head, 4)
    head = insert_head(head, 3)
    head = insert_head(head, 2)
    head = insert_head(head, 1)

    # ----- Insert 6 at the end -----
    tail = head
    while tail is not None and tail.next is not None:
        tail = tail.next
    insert_after(tail, 6)

    print_list(head)   # 1 2 3 4 5 6

    # ----- Delete 6 -----
    delete_after(tail)

    print_list(head)   # 1 2 3 4 5

    # ----- Reverse the list -----
    head = reverse_list(head)

    print_list(head)   # 5 4 3 2 1

    # ----- Check for cycle -----
    print(f"\nHas cycle: {'true' if has_cycle(head) else 'false'}")

    # ----- Build separate cycle list: 1 2 3 4 -----
    cycle: Optional[ListNode] = None
    cycle = insert_head(cycle, 4)
    cycle = insert_head(cycle, 3)
    cycle = insert_head(cycle, 2)
    cycle = insert_head(cycle, 1)

    print_list(cycle)  # 1 2 3 4

    # Create cycle: 4 -> 3
    p = cycle.next.next      # type: ignore[assignment]  # node 3
    q = p.next               # node 4
    q.next = p               # 4 -> 3

    print(f"\nHas cycle: {'true' if has_cycle(cycle) else 'false'}")

    # ----- Reverse head again -----
    head = reverse_list(head)   # 5 4 3 2 1 -> 1 2 3 4 5

    print_list(head)            # 1 2 3 4 5

    # ----- Middle node -----
    mid = middle_node(head)
    if mid is not None:
        print(f"\nMiddle node value: {mid.val}")

    # ----- Build second list: 6 7 8 -----
    more: Optional[ListNode] = None
    more = insert_head(more, 8)
    more = insert_head(more, 7)
    more = insert_head(more, 6)

    print_list(more)   # 6 7 8

    # ----- Merge lists -----
    merged = merge_two_lists(head, more)

    print_list(merged) # 1 2 3 4 5 6 7 8

    # ----- Remove 7 (2nd from end) -----
    merged = remove_nth_from_end(merged, 2)

    print_list(merged) # 1 2 3 4 5 6 8

if __name__ == "__main__":
    main()

File: code_samples/section3/problems/section3.cs
#nullable enable

static ListNode InsertHead(ListNode? head, int x) {
    return new ListNode(x, head);
}

static void InsertAfter(ListNode? p, int x) {
    if (p == null) return;
    var node = new ListNode(x, p.Next);
    p.Next = node;
}

static void DeleteAfter(ListNode? p) {
    if (p == null || p.Next == null) return;
    var doomed = p.Next;
    p.Next = doomed.Next;
}

static ListNode? ReverseList(ListNode? head) {
    ListNode? prev = null;
    ListNode? curr = head;

    while (curr != null) {
        var nxt = curr.Next;
        curr.Next = prev;
        prev = curr;
        curr = nxt;
    }
    return prev;
}

static bool HasCycle(ListNode? head) {
    if (head == null) return false;
    var slow = head;
    var fast = head;
    while (fast != null && fast.Next != null) {
        slow = slow!.Next!;
        fast = fast.Next.Next!;
        if (slow == fast) return true;
    }
    return false;
}

static ListNode? MiddleNode(ListNode? head) {
    var slow = head;
    var fast = head;

    while (fast != null && fast.Next != null) {
        slow = slow!.Next;
        fast = fast.Next.Next;
    }
    return slow;
}

static ListNode? MergeTwoLists(ListNode? l1, ListNode? l2) {
    var dummy = new ListNode(0);
    var tail = dummy;

    var a = l1;
    var b = l2;
    while (a != null && b != null) {
        if (a.Val <= b.Val) {
            tail.Next = a;
            a = a.Next;
        } else {
            tail.Next = b;
            b = b.Next;
        }
        tail = tail.Next!;
    }
    if (a != null) tail.Next = a;
    if (b != null) tail.Next = b;

    return dummy.Next;
}

static ListNode? RemoveNthFromEnd(ListNode? head, int n) {
    var dummy = new ListNode(0) { Next = head };
    var fast = dummy;
    var slow = dummy;

    for (int i = 0; i <= n; i++) {
        if (fast == null) return head; // n too large
        fast = fast.Next!;
    }

    while (fast != null) {
        fast = fast.Next!;
        slow = slow.Next!;
    }

    var doomed = slow.Next;
    if (doomed != null) {
        slow.Next = doomed.Next;
    }
    return dummy.Next;
}

static void PrintList(ListNode? head) {
    var current = head;
    Console.WriteLine();
    while (current != null) {
        Console.Write(current.Val);
        current = current.Next;
    }
}

// ----- Build list: 1 2 3 4 5 -----
ListNode? head = null;
head = InsertHead(head, 5);
head = InsertHead(head, 4);
head = InsertHead(head, 3);
head = InsertHead(head, 2);
head = InsertHead(head, 1);

// ----- Insert 6 at the end -----
ListNode tail = head!;
while (tail.Next != null) {
    tail = tail.Next;
}
InsertAfter(tail, 6);

PrintList(head);   // 123456

// ----- Delete 6 -----
DeleteAfter(tail);

PrintList(head);   // 12345

// ----- Reverse the list -----
head = ReverseList(head);

PrintList(head);   // 54321

// ----- Check for cycle -----
Console.WriteLine($"\nHas cycle: {(HasCycle(head) ? "true" : "false")}");

// ----- Build separate cycle list: 1 2 3 4 -----
ListNode? cycle = null;
cycle = InsertHead(cycle, 4);
cycle = InsertHead(cycle, 3);
cycle = InsertHead(cycle, 2);
cycle = InsertHead(cycle, 1);

PrintList(cycle);  // 1234

// Create cycle: 4 -> 3
ListNode p = cycle!.Next!.Next!;       // node 3
ListNode q = cycle.Next!.Next!.Next!;    // node 4
q.Next = p;

Console.WriteLine($"\nHas cycle: {(HasCycle(cycle) ? "true" : "false")}");

// ----- Reverse head again -----
head = ReverseList(head);   // 54321 -> 12345

PrintList(head);            // 12345

// ----- Middle node -----
ListNode? mid = MiddleNode(head);
if (mid != null) {
    Console.WriteLine($"\nMiddle node value: {mid.Val}");
}

// ----- Build second list: 6 7 8 -----
ListNode? more = null;
more = InsertHead(more, 8);
more = InsertHead(more, 7);
more = InsertHead(more, 6);

PrintList(more);            // 678

// ----- Merge lists -----
ListNode? merged = MergeTwoLists(head, more);

PrintList(merged);          // 12345678

// ----- Remove 7 (2nd from end) -----
merged = RemoveNthFromEnd(merged, 2);

PrintList(merged);          // 1234568

class ListNode(int val, ListNode? next = null)
{
    public int Val = val;
    public ListNode? Next = next;
}
